File: lib/native/pact_verifier.ex
defmodule Pact.Native.PactVerifier do
  @moduledoc false
  version = Mix.Project.config()[:version]

  use RustlerPrecompiled,
    otp_app: :pact_verifier_ex,
    crate: "pact_verifier_nif",
    base_url: "https://github.com/valerio-iachini/pact_verifier_ex/releases/download/v#{version}",
    force_build: System.get_env("PACT_CONSUMER_EX_FORCE_BUILD") == "true",
    version: version,
    targets: [
      "aarch64-apple-darwin",
      "aarch64-unknown-linux-gnu",
      "arm-unknown-linux-gnueabihf",
      "riscv64gc-unknown-linux-gnu",
      "x86_64-apple-darwin",
      "x86_64-pc-windows-gnu",
      "x86_64-pc-windows-msvc",
      "x86_64-unknown-linux-gnu"
    ]
    defmodule ProviderInfo do
      @moduledoc """
      ProviderInfo
      """
      @enforce_keys [:name, :protocol, :host, :port, :path, :transports]
      defstruct [
        :name,
        :protocol,
        :host,
        :port,
        :path,
        :transports
      ]

      @type t :: %__MODULE__{
        name: String.t(),
        protocol: String.t(),
        host: String.t(),
        port: nil | integer(),
        path: String.t(),
        transports: [ProviderTransport.t()]
      }
    end

    defmodule ProviderTransport do
      @moduledoc """
      ProviderTransport
      """
      @enforce_keys [:transport]
      defstruct [
        :transport,
        :port,
        :path,
        :scheme
      ]

      @type t :: %__MODULE__{
        transport: String.t(),
        port: nil | integer(),
        path: nil | String.t(),
        scheme: nil | String.t()
      }
    end

    defmodule HttpAuth do
      @moduledoc """
      HttpAuth
      """
      @type t ::
        {:user, String.t(), nil | String.t()} |
        {:token, String.t()} |
        :none
    end

    defmodule Link do
      @moduledoc """
      Link
      """
      @enforce_keys [:name, :templated]
      defstruct [
        :name,
        :href,
        :templated,
        :title
      ]

      @type t :: %__MODULE__{
        name: String.t(),
        href: nil | String.t(),
        templated: boolean(),
        title: nil | String.t()
      }
    end

    defmodule ConsumerVersionSelector do
      @moduledoc """
      ConsumerVersionSelector
      """
      defstruct [
        :consumer,
        :tag,
        :fallback_tag,
        :latest,
        :deployed_or_released,
        :deployed,
        :released,
        :environment,
        :main_branch,
        :branch,
        :matching_branch,
        :fallback_branch
      ]

      @type t :: %__MODULE__{
        consumer: nil | String.t(),
        tag: nil | String.t(),
        fallback_tag: nil | String.t(),
        latest: nil | boolean(),
        deployed_or_released: nil | boolean(),
        deployed: nil | boolean(),
        released: nil | boolean(),
        environment: nil | String.t(),
        main_branch: nil | boolean(),
        branch: nil | String.t(),
        matching_branch: nil | boolean(),
        fallback_branch: nil | String.t()
      }
    end

    defmodule PactSource do
      @moduledoc """
      PactSource
      """
      @type t ::
        :unknown |
        {:file, String.t()} |
        {:dir, String.t()} |
        {:url, String.t(), nil | HttpAuth.t()} |
        {:broker_url, String.t(), String.t(), nil | HttpAuth.t(), [Link.t()]} |
        {:broker_with_dynamic_configuration, %{
          provider_name: String.t(),
          broker_url: String.t(),
          enable_pending: boolean(),
          include_wip_pacts_since: nil | String.t(),
          provider_tags: [String.t()],
          provider_branch: nil | String.t(),
          selectors: [ConsumerVersionSelector.t()],
          auth: nil | HttpAuth.t(),
          links: [Link.t()]
        }} |
        {:string, String.t()} |
        {:webhook_callback_url, %{pact_url: String.t(), broker_url: String.t(), auth: nil | HttpAuth.t()}}
    end

    defmodule FilterById do
      @moduledoc """
      FilterById
      """
      @type t ::
        {:interaction_id, String.t()} |
        {:interaction_key, String.t()} |
        {:interaction_desc, String.t()}
    end

    defmodule FilterInfo do
      @moduledoc """
      FilterInfo
      """
      @type t ::
        :none |
        {:description, String.t()} |
        {:state, String.t()} |
        {:description_and_state, String.t(), String.t()} |
        {:interaction_ids, [FilterById.t()]}
    end

    defmodule PublishOptions do
      @moduledoc """
      PublishOptions
      """
      defstruct [
        :provider_version,
        :build_url,
        :provider_tags,
        :provider_branch
      ]

      @type t :: %__MODULE__{
        provider_version: nil | String.t(),
        build_url: nil | String.t(),
        provider_tags: [String.t()],
        provider_branch: nil | String.t()
      }
    end

    defmodule VerificationMetrics do
      @moduledoc """
      VerificationMetrics
      """
      defstruct [
        :test_framework,
        :app_name,
        :app_version
      ]

      @type t :: %__MODULE__{
        test_framework: String.t(),
        app_name: String.t(),
        app_version: String.t()
      }
    end

    defmodule HttpRequest do
      @moduledoc """
      HttpRequest
      """
      defstruct [
        :method,
        :path,
        :query,
        :headers
        # TODO: body, matching_rules, generators
      ]

      @type t :: %__MODULE__{
        method: String.t(),
        path: String.t(),
        query: nil | map(),
        headers: nil | map()
      }
    end

    defmodule RequestFilter do
      @moduledoc """
      RequestFilter
      """
      defstruct [
        :pid
      ]

      @type t :: %__MODULE__{
        pid: any()
      }
    end

    defmodule VerificationOptions do
      @moduledoc """
      VerificationOptions
      """
      defstruct [
        :request_filter,
        :disable_ssl_verification,
        :request_timeout,
        :custom_headers,
        :coloured_output,
        :no_pacts_is_error,
        :exit_on_first_failure,
        :run_last_failed_only
      ]

      @type t :: %__MODULE__{
        request_filter: nil | RequestFilter.t(),
        disable_ssl_verification: boolean(),
        request_timeout: integer(),
        custom_headers: map(),
        coloured_output: boolean(),
        no_pacts_is_error: boolean(),
        exit_on_first_failure: boolean(),
        run_last_failed_only: boolean()
      }
    end

    defmodule HttpRequestProviderStateExecutor do
      @moduledoc """
      HttpRequestProviderStateExecutor
      """
      defstruct [
        :state_change_url,
        :state_change_teardown,
        :state_change_body,
        :reties
      ]

      @type t :: %__MODULE__{
        state_change_url: nil | String.t(),
        state_change_teardown: boolean(),
        state_change_body: boolean(),
        reties: integer()
      }
    end

    @spec verify_provider(
        provider_info :: ProviderInfo.t(), 
        source :: list(PactSource.t()),
        filter :: FilterInfo.t(),
        consumers :: list(String.t()),
        verification_options :: VerificationOptions.t(),
        publish_options :: PublishOptions.t() | nil,
        provider_state_executor :: HttpRequestProviderStateExecutor.t(),
        metrics_data :: VerificationMetrics.t() | nil
    ) :: bool() 
    def verify_provider(_description, _interaction_type),
        do: :erlang.nif_error(:nif_not_loaded)
end

File: native/pact_verifier_nif/src/pact_verifier.rs
use std::{collections::HashMap, sync::Arc};

use pact_models::{
    bodies::OptionalBody,
    prelude::{Generators, HttpAuth, MatchingRules},
    v4::http_parts::HttpRequest,
};
use pact_verifier::{
    callback_executors::HttpRequestProviderStateExecutor, metrics::VerificationMetrics,
    pact_broker::Link, ConsumerVersionSelector, FilterById, FilterInfo, NullRequestFilterExecutor,
    PactSource, ProviderInfo, ProviderTransport, PublishOptions, VerificationOptions,
};
use rustler::{NifResult, NifStruct, NifTaggedEnum};

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.ProviderInfo"]
pub struct ExProviderInfo {
    pub name: String,
    #[deprecated(note = "Use transports instead")]
    pub protocol: String,
    pub host: String,
    #[deprecated(note = "Use transports instead")]
    pub port: Option<u16>,
    #[deprecated(note = "Use transports instead")]
    pub path: String,
    pub transports: Vec<ExProviderTransport>,
}

impl From<ExProviderInfo> for ProviderInfo {
    #![allow(deprecated)]
    fn from(value: ExProviderInfo) -> Self {
        Self {
            name: value.name,
            protocol: value.protocol,
            host: value.host,
            port: value.port,
            path: value.path,
            transports: value.transports.into_iter().map(Into::into).collect(),
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.ProviderTransport"]
pub struct ExProviderTransport {
    pub transport: String,
    pub port: Option<u16>,
    pub path: Option<String>,
    pub scheme: Option<String>,
}

impl From<ExProviderTransport> for ProviderTransport {
    fn from(value: ExProviderTransport) -> Self {
        Self {
            transport: value.transport,
            port: value.port,
            path: value.path,
            scheme: value.scheme,
        }
    }
}

#[derive(NifTaggedEnum)]
pub enum ExHttpAuth {
    User(ExUserAuth),
    Token(ExTokenAuth),
    None,
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.HttpAuth.UserAuth"]
pub struct ExUserAuth {
    username: String,
    password: Option<String>,
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.HttpAuth.TokenAuth"]
pub struct ExTokenAuth {
    value: String,
}

impl From<ExHttpAuth> for HttpAuth {
    fn from(value: ExHttpAuth) -> Self {
        match value {
            ExHttpAuth::User(ExUserAuth { username, password }) => {
                HttpAuth::User(username, password)
            }
            ExHttpAuth::Token(ExTokenAuth { value }) => HttpAuth::Token(value),
            ExHttpAuth::None => HttpAuth::None,
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.Link"]
pub struct ExLink {
    pub name: String,
    pub href: Option<String>,
    pub templated: bool,
    pub title: Option<String>,
}

impl From<ExLink> for Link {
    fn from(value: ExLink) -> Self {
        Self {
            name: value.name,
            href: value.href,
            templated: value.templated,
            title: value.title,
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.ConsumerVersionSelector"]
pub struct ExConsumerVersionSelector {
    pub consumer: Option<String>,
    pub tag: Option<String>,
    pub fallback_tag: Option<String>,
    pub latest: Option<bool>,
    pub deployed_or_released: Option<bool>,
    pub deployed: Option<bool>,
    pub released: Option<bool>,
    pub environment: Option<String>,
    pub main_branch: Option<bool>,
    pub branch: Option<String>,
    pub matching_branch: Option<bool>,
    pub fallback_branch: Option<String>,
}

impl From<ExConsumerVersionSelector> for ConsumerVersionSelector {
    fn from(value: ExConsumerVersionSelector) -> Self {
        Self {
            consumer: value.consumer,
            tag: value.tag,
            fallback_tag: value.fallback_tag,
            latest: value.latest,
            deployed_or_released: value.deployed_or_released,
            deployed: value.deployed,
            released: value.released,
            environment: value.environment,
            main_branch: value.main_branch,
            branch: value.branch,
            matching_branch: value.matching_branch,
            fallback_branch: value.fallback_branch,
        }
    }
}

#[derive(NifTaggedEnum)]
pub enum ExPactSource {
    Unknown,
    File(String),
    Dir(String),
    Url(String, Option<ExHttpAuth>),
    BrokerUrl(String, String, Option<ExHttpAuth>, Vec<ExLink>),
    BrokerWithDynamicConfiguration {
        provider_name: String,
        broker_url: String,
        enable_pending: bool,
        include_wip_pacts_since: Option<String>,
        provider_tags: Vec<String>,
        provider_branch: Option<String>,
        selectors: Vec<ExConsumerVersionSelector>,
        auth: Option<ExHttpAuth>,
        links: Vec<ExLink>,
    },
    String(String),
    WebhookCallbackUrl {
        pact_url: String,
        broker_url: String,
        auth: Option<ExHttpAuth>,
    },
}

impl From<ExPactSource> for PactSource {
    fn from(value: ExPactSource) -> Self {
        match value {
            ExPactSource::Unknown => PactSource::Unknown,
            ExPactSource::File(s) => PactSource::File(s),
            ExPactSource::Dir(s) => PactSource::Dir(s),
            ExPactSource::Url(url, auth) => PactSource::URL(url, auth.map(Into::into)),
            ExPactSource::BrokerUrl(url, provider, auth, links) => PactSource::BrokerUrl(
                url,
                provider,
                auth.map(Into::into),
                links.into_iter().map(Into::into).collect(),
            ),
            ExPactSource::BrokerWithDynamicConfiguration {
                provider_name,
                broker_url,
                enable_pending,
                include_wip_pacts_since,
                provider_tags,
                provider_branch,
                selectors,
                auth,
                links,
            } => PactSource::BrokerWithDynamicConfiguration {
                provider_name,
                broker_url,
                enable_pending,
                include_wip_pacts_since,
                provider_tags,
                provider_branch,
                selectors: selectors.into_iter().map(Into::into).collect(),
                auth: auth.map(Into::into),
                links: links.into_iter().map(Into::into).collect(),
            },
            ExPactSource::String(s) => PactSource::String(s),
            ExPactSource::WebhookCallbackUrl {
                pact_url,
                broker_url,
                auth,
            } => PactSource::WebhookCallbackUrl {
                pact_url,
                broker_url,
                auth: auth.map(Into::into),
            },
        }
    }
}

#[allow(clippy::enum_variant_names)]
#[derive(NifTaggedEnum)]
pub enum ExFilterById {
    InteractionId(String),
    InteractionKey(String),
    InteractionDesc(String),
}

impl From<ExFilterById> for FilterById {
    fn from(value: ExFilterById) -> Self {
        match value {
            ExFilterById::InteractionId(id) => FilterById::InteractionId(id),
            ExFilterById::InteractionKey(key) => FilterById::InteractionKey(key),
            ExFilterById::InteractionDesc(desc) => FilterById::InteractionDesc(desc),
        }
    }
}

#[derive(NifTaggedEnum)]
pub enum ExFilterInfo {
    None,
    Description(String),
    State(String),
    DescriptionAndState(String, String),
    InteractionIds(Vec<ExFilterById>),
}

impl From<ExFilterInfo> for FilterInfo {
    fn from(value: ExFilterInfo) -> Self {
        match value {
            ExFilterInfo::None => FilterInfo::None,
            ExFilterInfo::Description(description) => FilterInfo::Description(description),
            ExFilterInfo::State(state) => FilterInfo::State(state),
            ExFilterInfo::DescriptionAndState(description, state) => {
                FilterInfo::DescriptionAndState(description, state)
            }
            ExFilterInfo::InteractionIds(ex_filter_by_ids) => {
                FilterInfo::InteractionIds(ex_filter_by_ids.into_iter().map(Into::into).collect())
            }
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.PublishOptions"]
pub struct ExPublishOptions {
    pub provider_version: Option<String>,
    pub build_url: Option<String>,
    pub provider_tags: Vec<String>,
    pub provider_branch: Option<String>,
}

impl From<ExPublishOptions> for PublishOptions {
    fn from(value: ExPublishOptions) -> Self {
        Self {
            provider_version: value.provider_version,
            build_url: value.build_url,
            provider_tags: value.provider_tags,
            provider_branch: value.provider_branch,
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.VerificationMetrics"]
pub struct ExVerificationMetrics {
    pub test_framework: String,
    pub app_name: String,
    pub app_version: String,
}

impl From<ExVerificationMetrics> for VerificationMetrics {
    fn from(value: ExVerificationMetrics) -> Self {
        Self {
            test_framework: value.test_framework,
            app_name: value.app_name,
            app_version: value.app_version,
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.HttpRequest"]
pub struct ExHttpRequest {
    pub method: String,
    pub path: String,
    pub query: Option<HashMap<String, Vec<Option<String>>>>,
    pub headers: Option<HashMap<String, Vec<String>>>,
    // TODO: implement following fields
    //    pub body: OptionalBody,
    //    pub matching_rules: MatchingRules,
    //    pub generators: Generators,
}

impl From<HttpRequest> for ExHttpRequest {
    fn from(value: HttpRequest) -> Self {
        Self {
            method: value.method,
            path: value.path,
            query: value.query,
            headers: value.headers,
        }
    }
}

impl From<ExHttpRequest> for HttpRequest {
    fn from(value: ExHttpRequest) -> Self {
        Self {
            method: value.method,
            path: value.path,
            query: value.query,
            headers: value.headers,
            body: OptionalBody::Empty,
            // TODO: implment the following field
            matching_rules: MatchingRules {
                rules: HashMap::new(),
            },
            generators: Generators::default(),
        }
    }
}

// TODO
//#[derive(NifStruct)]
//#[module = "Pact.PactVerifier.RequestFilter"]
//pub struct ExRequestFilter {
//    pid: LocalPid,
//}
//
//impl Debug for ExRequestFilter {
//    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
//        f.debug_struct("ExRequestFilter").finish()
//    }
//}
//
//impl RequestFilterExecutor for ExRequestFilter {
//    fn call(
//        self: std::sync::Arc<Self>,
//        request: &pact_models::v4::http_parts::HttpRequest,
//    ) -> pact_models::v4::http_parts::HttpRequest {
//        let (_tx, rx) = mpsc::channel::<ExHttpRequest>();
//        let mut msg_env = OwnedEnv::new();
//        let _ = msg_env.send_and_clear(&self.pid, |env| {
//            (
//                rustler::Atom::from_str(env, "request").unwrap(),
//                env.make_ref(),
//                Into::<ExHttpRequest>::into(request.clone()),
//            )
//        });
//
//        match rx.recv() {
//            Ok(filtered_request) => Into::<HttpRequest>::into(filtered_request),
//            Err(_) => request.clone(), // fallback
//        }
//    }
//
//    fn call_non_http(
//        &self,
//        request_body: &pact_models::prelude::OptionalBody,
//        metadata: &HashMap<String, Either<serde_json::Value, Bytes>>,
//    ) -> (
//        pact_models::prelude::OptionalBody,
//        HashMap<String, Either<serde_json::Value, Bytes>>,
//    ) {
//        todo!()
//    }
//}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.VerificationOptions"]
pub struct ExVerificationOptions {
    //pub request_filter: Option<ExRequestFilter>,
    pub disable_ssl_verification: bool,
    pub request_timeout: u64,
    pub custom_headers: HashMap<String, String>,
    pub coloured_output: bool,
    pub no_pacts_is_error: bool,
    pub exit_on_first_failure: bool,
    pub run_last_failed_only: bool,
}

impl From<ExVerificationOptions> for VerificationOptions<NullRequestFilterExecutor> {
    fn from(value: ExVerificationOptions) -> Self {
        Self {
            request_filter: None,
            disable_ssl_verification: value.disable_ssl_verification,
            request_timeout: value.request_timeout,
            custom_headers: value.custom_headers,
            coloured_output: value.coloured_output,
            no_pacts_is_error: value.no_pacts_is_error,
            exit_on_first_failure: value.exit_on_first_failure,
            run_last_failed_only: value.run_last_failed_only,
        }
    }
}

#[derive(NifStruct)]
#[module = "Pact.PactVerifier.HttpRequestProviderStateExecutor"]
pub struct ExHttpRequestProviderStateExecutor {
    pub state_change_url: Option<String>,
    pub state_change_teardown: bool,
    pub state_change_body: bool,
    pub reties: u8,
}

impl From<ExHttpRequestProviderStateExecutor> for HttpRequestProviderStateExecutor {
    fn from(value: ExHttpRequestProviderStateExecutor) -> Self {
        Self {
            state_change_url: value.state_change_url,
            state_change_teardown: value.state_change_teardown,
            state_change_body: value.state_change_body,
            reties: value.reties,
        }
    }
}

#[allow(clippy::too_many_arguments)]
#[rustler::nif(name = "verify_provider", schedule = "DirtyIo")]
pub fn verify_provider(
    provider_info: ExProviderInfo,
    source: Vec<ExPactSource>,
    filter: ExFilterInfo,
    consumers: Vec<String>,
    verification_options: ExVerificationOptions,
    publish_options: Option<ExPublishOptions>,
    provider_state_executor: ExHttpRequestProviderStateExecutor,
    metrics_data: Option<ExVerificationMetrics>,
) -> NifResult<bool> {
    pact_verifier::verify_provider(
        provider_info.into(),
        source.into_iter().map(Into::into).collect(),
        filter.into(),
        consumers,
        &verification_options.into(),
        publish_options.map(Into::into).as_ref(),
        &Arc::new(Into::<HttpRequestProviderStateExecutor>::into(
            provider_state_executor,
        )),
        metrics_data.map(Into::into),
    )
    .map_err(|e| {
        rustler::Error::RaiseTerm(Box::new((
            "invalid_interaction_builder_reference",
            format!("{e:?}"),
        )))
    })
}
